### PR TITLE
Support TimeSpan with RangeAttribute in Options validation source generator

### PR DESCRIPTION
**Pull request by tarekgh**
https://github.com/dotnet/runtime/issues/94119

### CODE REVIEW

**Comment by tarekgh** [comment]
This is a candidate to port to 8.0 servicing. 

CC @ericstj @eerhardt

File: src/libraries/Microsoft.Extensions.Options/gen/Emitter.cs [hunk 1]
```diff
@@ -374,12 +374,83 @@ public void EmitCompareAttribute(string modifier, string prefix, string classNam
...
         {
             OutGeneratedCodeAttribute();
 
             string qualifiedClassName = $"{prefix}{suffix}_{className}";
 
+            string initializationString = emitTimeSpanSupport ?
+            """
+                                if (OperandType == typeof(global::System.TimeSpan))
+                                {
+                                    if (!global::System.TimeSpan.TryParse((string)Minimum, culture, out global::System.TimeSpan timeSpanMinimum) ||
```
Comment (eiriktsarpalis) [inline src/libraries/Microsoft.Extensions.Options/gen/Emitter.cs:387]: Could this parsing happen at compile time, or do we have to honor local culture settings?

File: src/libraries/Microsoft.Extensions.Options/gen/Emitter.cs [hunk 2]
```diff
@@ -374,12 +374,83 @@ public void EmitCompareAttribute(string modifier, string prefix, string classNam
 """);
         }
 
-        public void EmitRangeAttribute(string modifier, string prefix, string className, string suffix)
+        public void EmitRangeAttribute(string modifier, string prefix, string className, string suffix, bool emitTimeSpanSupport)
```
Comment (eiriktsarpalis) [inline src/libraries/Microsoft.Extensions.Options/gen/Emitter.cs:377]: Why is this emitting timespans conditional out of curiosity?

File: src/libraries/Microsoft.Extensions.Options/gen/Emitter.cs [hunk 3]
```diff
@@ -414,19 +485,20 @@ public void EmitRangeAttribute(string modifier, string prefix, string className,
                 string.Format(global::System.Globalization.CultureInfo.CurrentCulture, GetValidationErrorMessage(), name, Minimum, Maximum);
         private bool NeedToConvertMinMax { get; }
         private bool Initialized { get; set; }
+        private const string c_minMaxError = "The minimum and maximum values must be set to valid values.";
```
Comment (eiriktsarpalis) [inline src/libraries/Microsoft.Extensions.Options/gen/Emitter.cs:488]: I would recommend extracting all error messages to a standalone source file.

File: src/libraries/Microsoft.Extensions.Options/gen/Parser.cs [hunk 4]
```diff
@@ -641,23 +652,25 @@ private void TrackRangeAttributeForSubstitution(AttributeData attribute, ITypeSy
...
-                typeSymbol = ((INamedTypeSymbol)typeSymbol).TypeArguments[0];
-            }
-
-            if (argumentSpecialType != SpecialType.None &&
-                OptionsSourceGenContext.IsConvertibleBasicType(typeSymbol) &&
-                (constructorParameters.Length != 3 || typeSymbol.SpecialType == argumentSpecialType)) // When type is provided as a parameter, it has to match the property type.
+            if (argumentType is not null)
             {
-                _optionsSourceGenContext.EnsureTrackingAttribute(attribute.AttributeClass!.Name, createValue: false, out _);
+                _optionsSourceGenContext.EnsureTrackingAttribute(attribute.AttributeClass!.Name, createValue: hasTimeSpanType, out _);
```
Comment (eiriktsarpalis) [inline src/libraries/Microsoft.Extensions.Options/gen/Parser.cs:673]: I guess this asnwers my previous question :-)

**Review by eiriktsarpalis — APPROVED** [review]
LGTM modulo one question I had.

**Comment by eerhardt** [comment]
I tested this fix with the https://github.com/dotnet/extensions/tree/main/src/Libraries/Microsoft.Extensions.Http.Resilience library (which uses the Options Validation source generator) and I'm able to successfully publish a Native AOT application using the library built with this fix with no AOT warnings.

Thanks, @tarekgh!

File: src/libraries/Microsoft.Extensions.Options/gen/Emitter.cs (hunk 1 above)
Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Options/gen/Emitter.cs:387]: The culture is one reason. The other reason is if I parse it at compile time will need to pass the parsed values to the attribute constructors (possibly as objects) which will need adding a new constructor there and will need to add more emitting code to generate such constructor parameter. I don't think this is worth doing it in servicing release.

File: src/libraries/Microsoft.Extensions.Options/gen/Emitter.cs (hunk 3 above)
Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Options/gen/Emitter.cs:488]: We can do that separately to reduce the changes we are porting to servicing.

**Pull request merged**